### ISSUE TITLE
Process locks up when receiving too much information

### ISSUE THREAD

**Issue by dmitrig01** [riemann/riemann#554]
I've noticed some odd behavior, that I can consistently reproduce:

when under stress (usually, other parts of the system are having issues, so they start sending tons of data to the Riemann process), I see the following symptoms:

- dashboard starts showing really funky data (i.e., non-linear time), then no data at all, e.g.:
  ![screen shot 2015-05-07 at 1 13 30 pm](https://cloud.githubusercontent.com/assets/10422/7618484/44bcae18-f967-11e4-8329-df75f4563077.png)
- host CPUs go up to 100% CPU usage (though oddly, when i run `top`, I can see the CPUs at 100% usage, but it shows no processes at high usage - very odd)
- existing TCP connections to Riemann stay open, though I don't know how far the data gets

Here is my riemann configuration: https://gist.github.com/dmitrig01/f6c127ef96bd6ae82b02

Let me know what other information I can provide :-)

**Comment by dmitrig01** [comment]
fyi, it looks like we are usually sending around 30000 metrics per second, running on AWS, r3.4xlarge, Ubuntu 14.04, OpenJDK 1.7. During peak times, I would guess that we go to at least 50000 metrics per second, possibly 100k, but I don't know – only just added monitoring for number of stats :-)

**Comment by dmitrig01** [comment]
Just got into this kind of state again.

Here's some more diagnostic information:

![screen shot 2015-05-14 at 12 54 32 pm](https://cloud.githubusercontent.com/assets/10422/7640520/898a943e-fa38-11e4-8a8e-4f7c20ad01c4.png)

some of these numbers look incredibly high – would be great to diagnose why.

also, from running a basic ifconfig, looks like we're receiving roughly 400mb of stats per minute.

I think the 30000 metrics per second is what's going to riemann's index – it receives fewer – I think on the order of 10k-20k.

**Comment by dmitrig01** [comment]
interestingly the documentation mentions `riemann netty execution-handler queue size`, but I don't see it on the query `service =~ "riemann %"`

**Comment by AeroNotix** [comment]
What kind of behaviour are you expecting with input like this?

**Comment by btobolaski** [comment]
Its fairly clear from your initial issue what the problem is. If the cpus are at 100%, your riemann server is overloaded and can't accept all of the events that it is being sent. There are two ways to approach this, either resize the server to get more cpus or decrease the amount of processing that you have riemann doing.

As for the missing queue size stat, I think I saw it mentioned in the irc that netty no longer exposes it. The documentation should probably be changed for that if its the case.

It does seem strange that the dashboard is showing things out of order but you should fix the cpu issues first.

**Comment by aphyr** [comment]
Yeah, riemann streams don't resort in time so when you get overloaded, could see skew between events. Graph probably doesn't know how to handle that.
As far as "I overloaded it and it stopped working" goes, wellllll, I can't really help much there. I suggest a JVM profiler. Probably particular streams that can be optimized for your use case.

**Comment by dmitrig01** [comment]
Hmm, interesting. Yeah, I suppose you're right – I'm overlooking the obvious. I'll try putting it on a bigger machine. I was about to ask about how to figure out how where the bottlenecks are. I'll take a look at JVM profiling – never done it before, but gotta have a first for everything :-)

**Comment by dmitrig01** [comment]
feel free to close this issue as a user error :-)

**Comment by aphyr** [comment]
I mean, if Riemann is *unduly* slow I'd consider that a bug, but it's hard for me to fix a general problem like "slow". We try to provide backpressure and clients should drop load--may be worth checking that your clients wait for acknowledgement, back off, etc.

**Comment by dmitrig01** [comment]
I'll see where I get with the profiling, and report back. The client (I'm using the nodejs client, with some modifications) is definitely not listening to backpressure – adding that is a great suggestion.

**Comment by dmitrig01** [comment]
Ha, that backpressure worked miracles:

![screen shot 2015-05-14 at 4 58 06 pm](https://cloud.githubusercontent.com/assets/10422/7644499/74a38df6-fa5a-11e4-8659-1bfd4f08d970.png)

(here you can see it in a bad state and self-recover!!!)

**Comment by faxm0dem** [comment]
@dmitrig01 it would be interesting to see the behaviour with a minimalistic riemann config like `(streams)`

**Comment by aphyr** [comment]
Wooooo! :)

**Comment by dmitrig01** [comment]
Hmm, so I'm still occasionally seeing this. I've since changed my client-side logic so it accumulates data points every second and then seconds over aggregates - so the number of metrics over a given time period should be fairly similar.

One potential culprit I identified was that I was sending data to influx, which backs up occasionally, so I put it in an `async-queue`. However, the error persisted; I managed to take an hprof dump here: https://gist.github.com/dmitrig01/37815dd20e2b4ce49a2f

I then pulled that part out, issued a HUP, and soon after took this hprof dump. https://gist.github.com/dmitrig01/37815dd20e2b4ce49a2f

From my cursory analysis, it appears that it's not actually doing much - just network polling. However, for some reason it's going up to 99% CPU on most/all CPUs (when I run top)...
